### DISCUSSION TITLE
Feeling my way into perso of a Visa applet and wanted a sanity check on approach and perhaps help to understand the result...

### DISCUSSION THREAD

**Discussion by marknimbuspay** [martinpaljak/GlobalPlatformPro#369] [Q&A] [answered]
... I am hoping to use 'gptool' to load the visa applet and then in a separate invocation add DGI's for the Visa applet.

First question first - is it actually viable to use as multiple invocations of gptool or is incorrect?

My current invocations of gptool are to :-

1. Install the Visa supplied StubServer applet that the primary applet depends on and to create the the AID I aim to storing DGI's for : 
`java -jar ~/projects/GlobalPlatformPro/tool/target/gp.jar -v -X --install /Users/mark/projects/VisaApplet/VSDC_292/Cap_Files/StubServer_180320.cap --applet A0000000031650 --create 315041592E5359532E4444463031
`
2. Install the Visa applet
`java -jar ~/projects/GlobalPlatformPro/tool/target/gp.jar -v -X --install /Users/mark/projects/VisaApplet/VSDC_292/Cap_Files/VSDC_292_200420.cap --applet A0000000031056 --create A0000000031010 --params 02 --privileges CardLock` 

These two steps look to complete fine, but when I then try to use -store-data-chunk using 

`java -jar ~/projects/GlobalPlatformPro/tool/target/gp.jar -v -d --connect A0000000031010 --mode CLR 
--store-data-chunk
80E200011B92071882022000940818030301100203009F100706011103000000 
--store-data-chunk
80E200021991041682027C00941008020200100203001006060018010201
--store-data-chunk
80E200044201023F703D57104761739001010010D1012201012345565F200D4558414D504C452F504552534F9F1F18303132333430303030303030303030353536303030303030`

Please note:- 
- I specified --mode CLR so the data passed to the javacard matches as closely as possible match the example flow given in the documentation - which has no MAC'ing shown.
-  I have started by only specifying in the first **three** examples data-chunks to try the storing of these DGIs
- Currently the first STORE_DATA fails with 'STORE DATA failed: 0x6A88 (Referenced data not found)'.

The sanity checks I am seeking to make here are :
- Is using three distinct invocations of gptool a valid approach?
- Are my parameters on each command sensible and correct for my stated steps?

If the above all seems fine to you the reader, then any guidance on helping me to understand my 0x6A88 result on the first STORE would be appreciated.

I have attached the --list output of the javacard before the --store-data-chunk attempt :- 
[list_output_before_STORE_attempt.txt](https://github.com/user-attachments/files/18079248/list_output_before_STORE_attempt.txt)

and the output from the --store-data-chuck command :-

[output_from_STORE-DATA-CHUNK_attempt.txt](https://github.com/user-attachments/files/18079298/output_from_STORE-DATA-CHUNK_attempt.txt)

The javacard I have chosen to use here is an 'NXP J3H145 Dual Interface Java card - 144k'.

If I have missed sharing something important for you then please say I will try and fill any gaps I need to.

**Comment by marknimbuspay** [comment, marked as answer]
Having slept on this overnight, I spent needed time assessing the parameters and the data being written to the card.  I found my error. In my store-data-chunk payload I was erroneously including the STORE_DATA command header.  This command and length detail was being added as well by gptool causing the card to see the STORE_DATA command as the DGI bytes and rejecting it.

For clarity, I was specifying :- 
--store-data-chunk **80E200011B**92071882022000940818030301100203009F100706011103000000 

and I should (of course) have been just specifying the data I wanted to store :- 

--store-data-chunk 92071882022000940818030301100203009F100706011103000000 

and having gptool wrap it perfectly for me.

So In summary, my approach of using separate invocations of gptool seems ok and now my parameters into it are correct, I am moving forward again.
